### PR TITLE
[BUG] Fix KotlinReflectionInternalError for value-class @PreviewParameter previews

### LINKED ISSUE #141: [BUG] Scanning/invoking a @Preview with a value class @PreviewParameter throws KotlinReflectionInternalError
**Describe the bug**
`ComposablePreviewScanner` crashes when it encounters a `@Preview` whose `@PreviewParameter` provider supplies a Kotlin **value class** (e.g. `PreviewParameterProvider<Dp>`). It fails in two places:

- `getPreviews()` — while building the preview's display name (`ProvideComposablePreview.methodParametersType`)
- `ComposablePreview.invoke()` — while invoking the composable

Both resolve the preview `Method` through kotlin-reflect (`Method.kotlinFunction` / `ValueClassAwareCaller`). For a value-class signature, `ValueClassAwareCaller.checkParametersSize` asserts that the Kotlin-descriptor arity equals the JVM method arity and throws:

```
kotlin.reflect.jvm.internal.KotlinReflectionInternalError:
Inconsistent number of parameters in the descriptor and Java reflection object: 1 != 3
```

The mismatch exists because the Compose compiler appends synthetic parameters (`Composer`, an `Int` `changed` mask, and — with defaults — an `Int` default mask) that are **not** in the Kotlin descriptor. kotlin-reflect only routes value-class callables through `ValueClassAwareCaller`, which is why non-value-class previews are unaffected — they never hit the arity assertion.

**To Reproduce**
1. Define a value-class preview parameter:
   ```kotlin
   class DpValueClassProvider : PreviewParameterProvider<Dp> {
       override val values = sequenceOf(32.dp, 64.dp)
   }

   @Preview
   @Composable
   fun ValueClassPreviewParameterPreview(
       @PreviewParameter(provider = DpValueClassProvider::class) size: Dp,
   ) { Text(size.toString()) }
   ```
2. Scan it: `AndroidComposablePreviewScanner().scanPackageTrees("…").getPreviews()`
3. `KotlinReflectionInternalError: Inconsistent number of parameters … 1 != 3` is thrown (both on `getPreviews()`/`toString()` and on `invoke()`).

**Expected behavior**
The preview is scanned and invoked/rendered like any other `@PreviewParameter` preview, unboxing the value-class argument.

**Environment (please complete the following information):**
 - JDK Version: 17
 - AGP Version: 8.11.1
 - Composable Preview Scanner Version: 0.9.2 (`master`)
 - Kotlin Version: 2.0.0
 - Compose Version: BOM 2026.04.01 (compose-runtime 1.11.0)

**Additional context**
Root cause is that the enumeration/invocation paths use kotlin-reflect, which enforces a descriptor↔JVM arity consistency it cannot satisfy for Compose-generated methods. `androidx.compose.runtime.reflect.ComposableMethod` (`java.lang.reflect`-based, Compose-ABI aware) already models the synthetic params and is the natural replacement. I have a fix + tests and will open a PR referencing this issue.

### PR DESCRIPTION
**Pull request by nshurtz**
Fixes #141

### Problem
A `@Preview` whose `@PreviewParameter` provider yields a Kotlin value class crashed both `getPreviews()` and `ComposablePreview.invoke()` with `KotlinReflectionInternalError: Inconsistent number of parameters … arity != expectedArgsSize`. The scanner resolved the preview method via kotlin-reflect, whose `ValueClassAwareCaller` asserts the Kotlin-descriptor arity equals the JVM arity — which never holds for `@Composable` functions, because the Compose compiler appends synthetic `Composer`/`changed`/default-mask parameters absent from the descriptor. Only value-class previews are routed through that caller, so only they crashed.

### Fix
Switch both paths off kotlin-reflect onto `androidx.compose.runtime.reflect.ComposableMethod` (`java.lang.reflect`-based, Compose-ABI aware):
- `ProvideComposablePreview.methodParametersType` — derive the real parameter count via `asComposableMethod()` instead of `kotlinFunction`.
- `ComposablePreviewInvocationHandler` — invoke via `ComposableMethod.invoke`, unboxing value-class arguments; preserves the `PreviewWrapper` path, bound-instance invocation, and the "`@PreviewParameter` must be first" guard.

### Tests
- New `PreviewParameterProvider<Dp>` fixture in its own top-level package (so the broad invoke suites don't pick it up).
- Enumeration/`toString` test asserting the value-class preview is scanned without crashing.
- Paparazzi render test exercising the invocation path; baselines recorded on Linux/JDK 17 to match CI.

### Notes
The scanned id keeps the raw JVM method name, which for a value-class parameter is mangled (`<name>-<hash>`); this is intentional so previews whose value-class params share an underlying JVM type don't collide (see the comment in `ComposablePreviewToStringTest`).

---
_This contribution (code and description) was prepared with AI assistance ([Claude Code](https://claude.com/claude-code)) and reviewed by me before submission._

## Summary by CodeRabbit

- **New Features**
  - Added support for Compose previews that use value-class parameters supplied by preview parameter providers.
  - Added compatibility for AndroidX and JetBrains Compose preview parameter annotations.
  - Improved handling of preview parameters, receivers, and default values during preview rendering.

- **Bug Fixes**
  - Prevented crashes when scanning and rendering previews with value-class parameters.

- **Tests**
  - Added coverage for preview discovery, stable preview identifiers, and Paparazzi and Roborazzi screenshot rendering.

### CODE REVIEW

**Comment by sergio-sastre** [comment]
@nshurtz Thanks for reporting the issue and opening a PR!
It looks great overall. I appreciate it! 😊

The only thing I am missing is to check that the changes work also in Roborazzi.

The new Composable sample that is used to reproduce the bug, seems that it is placed in a package that the Roborazzi tests do not scan. That’s also why a new Parameterized test was needed, the Roborazzi one is missing though.
The reason for this is, that I’ve seen in the past some code changes failing with only one of the Libraries due to the differences in how Paparazzi & Roborazzi download resources.

FYI, I am currently on holidays till the end of the week, but I promise to do the final check and release a bugfix version as soon as I am back. Sorry for the inconvenience! 🙏

File: core/src/main/java/sergio/sastre/composable/preview/scanner/core/preview/ComposablePreviewInvocationHandler.kt [hunk 1]
```diff
@@ -30,51 +29,97 @@ internal class ComposablePreviewInvocationHandler(
...
+            false -> composableMethod.declaringClass.getDeclaredConstructor()
+                .apply { isAccessible = true }
+                .newInstance()
+        }
+
+        return composable.invoke(
+            requireNotNull(composer) { "A Composer is required to invoke a @Composable preview" },
+            receiver,
+            *realArgs,
+        )
```
[the remark this answers is not in the document]
Comment (nshurtz) [reply core/src/main/java/sergio/sastre/composable/preview/scanner/core/preview/ComposablePreviewInvocationHandler.kt:88]: This is intentional, and we're keeping the current behavior. The scenario is a `@PreviewParameter` whose parameter *also* declares a Kotlin default and whose provider emits `null` — a degenerate combination (the provider normally supplies the value, so the default is otherwise dead).

The "null → default" outcome comes directly from androidx's own `ComposableMethod.invoke`, which is what we use for invocation: it derives the default mask from argument null-ness, so a `null` argument sets the default bit. Preserving the explicit `null` instead would require distinguishing "provider emitted null" from "argument absent" *and* knowing the parameter declares a default — and detecting a default on a value-class signature means `kotlin-reflect`, which throws `KotlinReflectionInternalError` on value classes (the exact crash this PR fixes). So a null-preserving path would reintroduce the fragility we're removing.

It's also not a regression: on `main` this combination crashed outright (`IllegalArgumentException: Callable expects 4 arguments, but 5 were provided`, for a plain `String?` too), and Android Studio's own preview renderer applies the default here as well — while crashing on the non-null value-class values this PR now renders. Leaving the invocation path as-is.

File: core/src/main/java/sergio/sastre/composable/preview/scanner/core/preview/ProvideComposablePreview.kt [hunk 2]
```diff
@@ -54,24 +54,24 @@ class ProvideComposablePreview<T> {
...
             private fun methodParametersTypeAsString(): String {
                 val previewMethod = composablePreviewMapper.previewMethod
-                val hasDefaultParams = previewMethod.kotlinFunction!!.parameters.any { it.isOptional }
-                val count = if (hasDefaultParams) 3 else 2
+                val realParametersCount = previewMethod.asComposableMethod()?.parameterCount
+                    ?: (previewMethod.parameterTypes.size - 2).coerceAtLeast(0)
                 return previewMethod
                     .genericParameterTypes
-                    .dropLast(count)
+                    .take(realParametersCount)
```
[the remark this answers is not in the document]
Comment (nshurtz) [reply core/src/main/java/sergio/sastre/composable/preview/scanner/core/preview/ProvideComposablePreview.kt:87]: Leaving this as-is. The `?:` fallback only runs when `asComposableMethod()` returns `null`, which happens only for a method that isn't a valid `@Composable` — but `previewMethod` here is always a scanned `@Composable` preview, so the primary path (`.parameterCount`, which correctly excludes the synthetic params) is what executes in practice. The fallback also only affects the screenshot-ID string, not rendering. Not worth changing for a branch that isn't reachable for real previews.

**Comment by nshurtz** [comment]
For context on the boundaries of this fix: while verifying, I found that Android Studio's own preview renderer (`androidx.compose.ui.tooling.ComposableInvoker`) doesn't handle nullable value-class `@PreviewParameter`s either — it throws `IllegalArgumentException: argument type mismatch` on them, because it passes the boxed value-class instance straight to `Method.invoke` without unboxing to the underlying type. This PR's invocation path renders them correctly.

I'm filing that (and the separate `@PreviewParameter`-with-default behavior) upstream with Google as Compose tooling issues; ~happy to link them here once opened~[Issue Tracker Link](https://issuetracker.google.com/issues/545636893). This PR stays scoped to the scanner-side crash fix.

**Comment by sergio-sastre** [comment]
> @nshurtz Thanks for reporting the issue and opening a PR!
> 
> It looks great overall. I appreciate it! 😊
> 
> 
> 
> The only thing I am missing is to check that the changes work also in Roborazzi.
> 
> 
> 
> The new Composable sample that is used to reproduce the bug, seems that it is placed in a package that the Roborazzi tests do not scan. That’s also why a new Parameterized test was needed, the Roborazzi one is missing though.
> 
> The reason for this is, that I’ve seen in the past some code changes failing with only one of the Libraries due to the differences in how Paparazzi & Roborazzi download resources.
> 
> 
> 
> FYI, I am currently on holidays till the end of the week, but I promise to do the final check and release a bugfix version as soon as I am back. Sorry for the inconvenience! 🙏

I just moved the tests to keep the current structure, ensure that the name includes the Value Class name instead of the one of the underlying class, and add some extra tests to cover several @PreviewParameter naming edge cases.

I'll release a bugfix version as soon as I find the time, likely today or tomorrow.

Thank you very much for your contribution!
